File: funciones.rb
# frozen_string_literal: true

# Funciones generales
# Version 0.12: ajusto fechas / horas / combo box / radio buttons
# Version 1.04: soluciono error en str_to_float
# Version 1.10: agrego funcion para poner valor a checkboxes IMPORTANTE

require 'date'
require 'bigdecimal'

# Devuelve enter sin comas ni puntos
def entero_pelado(num)
  if num.is_a? String
    num.replace(',', '') if num.include? ','
    num = num.to_f if num.include? '.'
  end
  num.to_i
end

def str_to_float(cadena)
  # Ojo: probable falla cuando es $ 1.234.567,89 -> solucion
  # usar GSUB en lugar de SUB
  # Elimina espacios
  cad2 = cadena.nil? ? '' : cadena.gsub(' ', '')
  # cadena = cadena.replace("-","")
  # Elimina signo de $
  cad2 = cad2.empty? ? '' : cad2.gsub('$', '')
  # Devuelve CERO is está vacío o solo tiene un guión
  return 0.00 if cad2 == '' || cad2.nil? || cad2 == '-'

  # Elimina paréntesis y sustituye por signo de menos
  if cad2[0] == '('
    cad2 = cad2.sub '(', ''
    cad2 = cad2.sub ')', ''
    cad2 = "-#{cad2}"
  end
  # Evaluar separador de decimlaes
  case cad2[-3..-3]
  when '.'
    # entonces remover coma como separador de miles
    cad2 = cad2.gsub ',', ''
  when ','
    # sacar punto como separador de miles
    cad2 = cad2.gsub '.', ''
    # luego poner punto como separador decimal
    cad2 = cad2.sub! ',', '.'
  end
  cad2.to_f
end

# Devuelve decimal sin separador de miles
def decimal_pelado(num)
  num.sub! ',', '' if num.is_a?(String) && num.include?(',')
  num.to_f
end

# transforma un numero en una cadena con formato moneda
def moneda(num)
  mon = if num.instance_of? BigDecimal
          num.to_s
        else
          numero(num, 2)
        end
  "$#{mon}"
end

# Reemplaza CADENA con REEMPLAZO en POSICION, sin alterar la longitud de la cadena
def substr_replace(cadena, reemplazo, posicion)
  cadena[:posicion] + reemplazo + cadena[posicion + len(reemplazo)..-1]
end

# convierte un flotante en decimal
def to_decimal(float_price)
  decimal.Decimal('%<float_price>.2f', float_price: float_price)
end

def to_boolean(variable)
  return False if ['False', 'false', 0, '0', nil].include? variable

  true
end

# Convierte a string, si es None queda cadena vacia
def nstr(item)
  return if item.nil?

  item.to_s
end

# Formatea un numero con decimales
def numero(num, places = 0)
  num = num.to_f if num.is_a? String
  num = num.to_f if num.is_a? BigDecimal
  num = 0 if num.nil?
  places = [0, places].max
  # "%.#{places}f" % num.round(places)
  # sprintf("%#{places}f", num.round(places))
  format("%.#{places}f", num.round(places))
end

# barra de progreso
def progreso(cuenta, total, estado = '')
  largo = 50
  largo_lleno = round(largo * cuenta / total.to_f).to_i
  porcentaje = round(100.0 * cuenta / total.to_f, 1)
  barra = '=' * largo_lleno + '-' * (largo - largo_lleno)
  print "[#{barra}] #{porcentaje}% ...#{estado}"
  $stdout.flush
end

# Cargar datos de un formulario a un registro active_record
def cargar_datos(formulario, registro, indice)
  formulario.each do |llave, valor|
    next if llave == indice

    registro.send("#{llave}=", valor)
  end
end

def valor_checkbox(parametro)
  parametro.nil? ? nil : parametro.to_i
end

File: funciones.py
"""Funciones generales
 Ver 0.10.28.
 Ver 2.01: agrego cargar datos  (x peewee), valor de checkbox
 Ver 2.10: soluciono perid "Sin datos" para nombre_cliente
 TODO: sacar nombre_cliente para otro lado.
 """

import sys
import decimal
from modelos import db_magik


def entero_pelado(num):
    """Devuelve entero sin comas ni puntos"""
    if isinstance(num, str):
        if "," in num:
            num = num.replace(",", "")
        if "." in num:
            num = float(num)
    return int(num)


def str_to_float(cadena: str) -> float:
    """Convierte cadena a flotante"""
    cadena = cadena.replace(" ", "")
    # cadena = cadena.replace("-","")
    cadena = cadena.replace("$", "")
    if cadena == '' or cadena is None or cadena == "-":
        return 0.00
    if cadena[0] == '(':
        cadena = cadena.replace("(", "")
        cadena = cadena.replace(")", "")
        cadena = "-" + cadena
    if cadena[-3:-2] == '.':
        # entonces remover coma como separador de miles
        cadena = cadena.replace(",", "")
    elif cadena[-3:-2] == ',':
        # sacar punto como separador de miles
        cadena = cadena.replace(".", "")
        # luego poner punto como separador decimal
        cadena = cadena.replace(",", ".")
    return float(cadena)


def decimal_pelado(num: str) -> float:
    """Devuelve decimal sin separador de miles"""
    if num is None:
        num = 0
    if "," in num:
        num = num.replace(",", "")
    return float(num)


def substr_replace(cadena, reemplazo, posicion):
    """Reemplaza CADENA con REEMPLAZO en POSICION, sin alterar la longitud de la cadena"""
    cad = cadena[:posicion] + reemplazo + cadena[posicion + len(reemplazo):]
    return cad


def to_decimal(float_price):
    """convierte un flotante en decimal"""
    return decimal.Decimal('%.2f' % float_price)


def to_boolean(variable) -> bool:
    if variable in ['False', 'false', 0, '0', None]:
        return False
    else:
        return True


def nstr(item):
    """Convierte a string, si es None queda cadena vacía"""
    # if isinstance(item, type(None)):
    if item is None:
        return ''
    return str(item)


def progreso(cuenta, total, estado=''):
    """barra de progreso"""
    largo = 50
    largo_lleno = int(round(largo * cuenta / float(total)))
    porcentaje = round(100.0 * cuenta / float(total), 1)
    barra = '=' * largo_lleno + '-' * (largo - largo_lleno)
    sys.stdout.write('[%s] %s%s ...%s\r' % (barra, porcentaje, '%', estado))
    sys.stdout.flush()


def cargar_datos(formulario, registro, indice):
    """carga datos a la tabla a partir de un formulario web (bottle)"""
    for clave, valor in formulario.items():
        if clave == indice:
            continue
        # Si no lo hago así, no registra UNICODE
        setattr(registro, clave, getattr(formulario, clave))


def valor_checkbox(parametro):
    try:
        retorno = int(parametro)
    except ValueError:
        retorno = 0
    return retorno


def entero(valor: str) -> int:
    """ devuelve entero, para evitar error al convertir INT"""
    if valor == "":
        return 0
    if valor is None:
        return 0
    try:
        retorno = int(float(valor))
    except ValueError:
        retorno = 0
    return retorno


def nombre_cliente(tipo: int, ident: int) -> str:
    if tipo == 1:
        if ident == 1117:
            nombre = 'CAMCEL'
        else:
            empr = db_magik.Empresas.get_or_none(ident)
            nombre = empr.emprdes if empr else "Sin datos"
    elif tipo == 2:
        cont = db_magik.Contratos.get_or_none(ident)
        if cont:
            pers = db_magik.Personas.get_or_none(cont.perid)
            nombre = pers.nombre() if pers else "Sin datos"
        else:
            nombre = "Sin datos"
    elif tipo == 3:
        pers = db_magik.Personas.get_or_none(ident)
        nombre = pers.nombre() if pers else "Sin datos"
    elif tipo == 4:
        nombre = "Solicitado por UNEM"
    else:
        nombre = "Sin datos"
    return nombre
